pymatgen/transformations/advanced_transformations.py

```diff
@@ -1682,31 +1682,23 @@ def apply_transformation(self, structure: Structure) -> Structure:
         sc_not_found = True
 
         if self.force_diagonal:
-            # trans_mat_diagonal holds the diagonal of the trans_mat
-            trans_mat_diagonal = np.array([0, 0, 0])
-            trans_mat_diagonal_update = np.array([1, 1, 1])
-        else:
-            # target_threshold is used as the desired cubic side lengths
-            target_sc_size = self.min_length
+            scale = self.min_length / np.array(structure.lattice.abc)
+            self.transformation_matrix = np.diag(np.ceil(scale).astype(int))
+            st = SupercellTransformation(self.transformation_matrix)
+            return st.apply_transformation(structure)
 
+        # target_threshold is used as the desired cubic side lengths
+        target_sc_size = self.min_length
         while sc_not_found:
-            if self.force_diagonal:
-                # Update trans_mat (with diagonal constraint)
-                trans_mat_diagonal += trans_mat_diagonal_update
-                self.transformation_matrix = np.diag(trans_mat_diagonal)
-
-            else:
-                # Update trans_mat (without diagonal constraint)
-                target_sc_lat_vecs = np.eye(3, 3) * target_sc_size
-
-                self.transformation_matrix = (
-                        np.linalg.inv(lat_vecs) @ target_sc_lat_vecs
-                )
+            target_sc_lat_vecs = np.eye(3, 3) * target_sc_size
```

Comment:
@utf if you want a square supercell where the _c_ vector is left untouched, then you need the last/third row of the transformation matrix to be fixed to [0, 0, 1]. So in that case, I would try adding something along the lines of `if want_square_supercell: target_sc_lat_vecs[-1,-1] = 1` right after `target_sc_lat_vecs = np.eye(3, 3) * target_sc_size`. Hope that helps.

Reply:
Thanks @rees-c, I'd be happy to add this in another PR (so as not to create merge conflicts with this one)

Reply:
@rees-c, making this change I have issues with the transformation matrix being singular, do you have any suggestions?

Reply:
To share an example, I'm getting a matrix like:

```
[[-2.83564613  2.83564613  2.83564613]
 [ 2.83564613 -2.83564613  2.83564613]
 [ 0.18413287  0.18413287 -0.18413287]]
```

which is

```
[[-3,  3,  3],
[ 3, -3,  3],
[ 0,  0, -1]]
```

after `_round_and_make_arr_singular`, but is still singular.

Reply:
Ah yikes, sorry about that. It looks like I only considered zero rows and columns when writing `_round_and_make_arr_singular`. Let me think about it and get back to you.

Reply:
I've figured out what's happening and how to fix it. For documentation's sake, I'll explain here and then add a suggestion later with the code fix. So the goal of `_round_and_make_arr_singular` is to round a matrix's entries to the nearest integers, and then make as small changes as possible to make the array singular. So in your example, my algorithm would have rounded the original matrix to
`[[-3, 3, 3],`  
`[3, -3, 3],`  
`[0, 0, 0]]`,
identified the third row as zero, and then chosen the entry in that row corresponding with the largest magnitude entry of the original row, `[0.18413287  0.18413287 -0.18413287]`. Since every entry of this original row have the same magnitude, my code broke ties randomly and in your case, just so happened to pick the one out of the three choices that doesn't make the matrix singular. I'll add a suggestion that fixes this by trying every entry instead of just randomly picking one. I guess it might be possible that trying every entry would still not work, but it should be incredibly unlikely (singular square matrices apparently are an infinitely thin subset of all square matrices). If by divine intervention the matrix is still not singular, I would then try all pairs of entries in the zero row, then all the entries in the zero row, then repeat this process (of changing singlets, pairs, triplets, etc.) for zero entries in the full matrix, and then just start picking random entries in the full matrix.

Reply:
Thanks for this exposition @rees-c, much appreciated in clearing it up!

To clarify, I'm assuming you're talking about this line:

https://github.com/materialsproject/pymatgen/blob/ce35198c83d4541cab36bfe4b12ef0631c8ec89c/pymatgen/transformations/advanced_transformations.py#L1915

And suggesting we try all values there along with an explicit check that the matrix isn't singular.

It seems like this bit of logic is only implemented for the rows, not for the columns, why is that?

Reply:
I have a test implementation up here: https://github.com/materialsproject/pymatgen/compare/square-transformation

It basically works, in that `|a| == |b| != |c|`, but angles are 45º, 90º, 90º -- the transformation I was aiming for had angles 90º, 90º, 90º with `|a| == |b| != |c|`.

Reply:
Hi @mkhorton! I see now that my code rounded every entry of each zero column instead of picking a random entry... if there was a good reason for why I would've done that, it has since escaped my brain haha.

With regard to the angles, it's my understanding that's it's not always possible to get perfect 90 degree angles (due to the requirement for all entries of the transformation to be integers for preserving periodicity). Not sure if that's why you're getting a 45 degree angle but I'll take a closer look at your test implementation and let you know if I spot anything.